resources/assets/lib/beatmap-discussions/editor-insertion-menu.tsx

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import BeatmapJsonExtended from 'interfaces/beatmap-json-extended';
+import * as _ from 'lodash';
+import { Portal } from 'portal';
+import * as React from 'react';
+import { Editor as SlateEditor, Element as SlateElement, Node as SlateNode, Point, Transforms } from 'slate';
+import { ReactEditor } from 'slate-react';
+import { SlateContext } from './slate-context';
+
+const editorClass = 'beatmap-discussion-editor';
+
+interface Props {
+  currentBeatmap: BeatmapJsonExtended;
+}
+
+export class EditorInsertionMenu extends React.Component<Props> {
+  static contextType = SlateContext;
+  bn = 'beatmap-discussion-editor-insertion-menu';
+  hideInsertMenuTimer?: number;
+  hoveredBlock: HTMLElement | undefined;
+  insertRef: React.RefObject<HTMLDivElement>;
+  menuPos: string | undefined;
+  mouseOver = false;
+  scrollContainer: HTMLElement | undefined;
+  // setTimeout delay is to prevent flashing when hovering the menu (the portal is not inside the container, so it fires a mouseleave)
+  throttledContainerMouseExit = _.throttle(() => { setTimeout(this.hideMenu.bind(this), 100); }, 10);
+  throttledContainerMouseMove = _.throttle(this.containerMouseMove.bind(this), 10);
+  throttledMenuMouseEnter = _.throttle(this.menuMouseEnter.bind(this), 10);
+  throttledMenuMouseExit = _.throttle(this.menuMouseLeave.bind(this), 10);
+  throttledScroll = _.throttle(this.forceHideMenu.bind(this), 10);
+
+  constructor(props: Props) {
+    super(props);
+
+    this.insertRef = React.createRef();
+  }
+
+  componentDidMount() {
+    if (this.insertRef.current) {
+      $(this.insertRef.current).on('mouseenter.editor-insert-menu', this.throttledMenuMouseEnter);
+      $(this.insertRef.current).on('mouseleave.editor-insert-menu', this.throttledMenuMouseExit);
+    }
+    $(window).on('scroll.editor-insert-menu', this.throttledScroll);
+  }
+
+  // updates cascade from our parent (slate editor), i.e. `componentDidUpdate` gets called on editor changes (typing/selection changes/etc)
+  componentDidUpdate() {
+    this.forceHideMenu();
+  }
+
+  componentWillUnmount() {
+    $(window).off('.editor-insert-menu');
+    if (this.scrollContainer) {
+      $(this.scrollContainer).off('.editor-insert-menu');
+    }
+    if (this.insertRef.current) {
+      $(this.insertRef.current).off('.editor-insert-menu');
+    }
+  }
+
+  containerMouseMove(event: JQuery.MouseMoveEvent) {
+    const block = event.target.closest(`.${editorClass}__block`);
```

Comment:
not sure about this... changing the target class would involve changing it in the parent `Editor`, leaking implementation outside of the (otherwise) self-contained component?

Reply:
?_? but it's already depending on the parent with specific styling class name to exist

Reply:
To put it another way: because this is an add-on to `Editor` (and not the other way around), `Editor` shouldn't require changes for this to work IMO

(also high chance of the `.js-x` class being left behind if/when this menu is changed/replaced)

That's the logic behind my thinking anyway 🤷‍♂️

Reply:
Editor already needs some changes for this to work though :thinking: 

Also I'm thinking it more like providing explicit interface with `.js-` (and this component is just consuming it) whereas with styling class name it's like depending on specific internal structure.

I've broke several things when there's no clear indication something is depended by something else :man_dancing: 

(it already depends on another internal structure - like specific element hierarchy - but that's another thing to worry about)